### PR TITLE
storage: calculate last_change_ts in acquire_pessimistic_lock

### LINKED ISSUE #13694: Skip accumulated locks by a second get/seek
## Motivation
Consider a use case: the user application always reads and locks a fixed key but never changes it:

```rust
txn = client.start_transaction()
txn.lock("k1")
v1 = txn.get("k1")
txn.put("k2", "v2")
txn.commit()
```

No matter it's an optimistic or pessimistic transaction, the transaction will write a new `LOCK` record on `k1` in the write CF. If we repeats executing this transaction, there can be a lot of `LOCK` records on `k1` that are newer than the effective `PUT` record, like:

```
k1@100  LOCK
k1@99   LOCK
k1@98   LOCK
...
k1@11   LOCK
k1@10   PUT   value = "v1"
```

It will be a disaster when we want to read the value of `k1`. If it's a point get, we can only seek to the latest version and waste a lot of effort calling next until we find an effective record. The performance of range scan is also degraded. Furthermore, the current MVCC GC cannot remove these `LOCK` records because they are newer than the latest `PUT` records. So, the performance will keep getting worse until the key is effectively modified.

## Brief design

One of the conditions of the problem is that we are seeking the latest version of the key. If we can know what is the version of the latest `PUT` or `DELETE` record, we can directly move the cursor to that key without skipping the huge amount of ineffective versions.

To achieve this, we will store two more fields in the `LOCK` records in the write CF:

```rust
pub struct Write {
    ...
    /// The commit TS of the latest PUT/DELETE record
    last_change_ts: TimeStamp,
    /// The number of versions that need skipping from this key
    /// to find the latest PUT/DELETE record
    versions_to_last_change: u64,
}
```

If we find that the latest version of a key is a LOCK, we may choose to get or seek again according to the `last_change_ts`.
Because get/seek is a much heavier operation than next, `versions_to_last_change` will tell us whether it is worth using get/seek.

Refer to https://pingcap.feishu.cn/docx/IwomdGtc1oIVepx9tIec2Aycn7c for more detailed design. There are also some discussions in the doc.

## Development tasks

- [x] https://github.com/pingcap/kvproto/pull/997
- [x] #13698
- [x] #13728
- [x] #13721
- [x] #13717
- [x] ~~#13749~~ (reverted by #13834)
- [x] #13759
- [x] #13784
- [x] #13819
- [x] #13820

## Further work

- Now, we don't calculate `last_change_ts` for a newly written record if the latest record does not contain `last_change_ts`. If the cluster is upgraded from old version, the enhancement may not work. We probably need to solve the problem but this may introduce unexpected extra work after upgrading.

### PR DESCRIPTION
**Pull request by sticnarf**
### What is changed and how it works?

Issue Number: Ref #13694

What's Changed:

```commit-message
Information about the last change is calculated and stored in the
pessimistic lock.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by ekexium — APPROVED** [review]
This pr should not close the linked issue

**Comment by sticnarf** [comment]
> This pr should not close the linked issue

Thanks for reminding. Changed to ref.

**Review by youjiali1995 — APPROVED** [review]

File: components/txn_types/src/write.rs [hunk 1]
```diff
@@ -259,6 +259,25 @@ impl Write {
...
+            WriteType::Put | WriteType::Delete => (commit_ts, 1),
+            WriteType::Lock | WriteType::Rollback => {
+                // If `last_change_ts` is zero, do not set `last_change_ts` to indicate we don't
+                // know where is the last change.
+                // This should not happen if data is written in new version TiKV. If we hope to
+                // support data from old TiKV, consider iterating to the last change to find it.
+                if !self.last_change_ts.is_zero() {
+                    (self.last_change_ts, self.versions_to_last_change + 1)
+                } else {
+                    (TimeStamp::zero(), 0)
```
Comment (MyonKeminta) [inline components/txn_types/src/write.rs:276]: Is it possible that we just return `(commit_ts, 1)` in this case, and if we seek to the `last_change_ts` and find it's Lock or Rollback, do `next`s to find the actual version? I think in this way maybe it's possible that for those keys that are frequently locked but rarely updated, they can benefit from the optimization after upgrading from an old version.

By the way, GC may change the real number of versions to last change. Does that matter?

Comment (sticnarf) [reply components/txn_types/src/write.rs:276]: > Is it possible that we just return `(commit_ts, 1)` in this case, and if we seek to the `last_change_ts` and find it's Lock or Rollback when reading, do `next`s to find the actual data? I think in this way maybe it's possible that for those keys that are frequently locked but rarely updated, they can benefit from the optimization after upgrading from an old version.

Emm, actually I don't intend to solve the problem for upgraded data at the first step.

If we indeed want to solve the problem for upgraded version, my idea is to iterate until the last change in `acquire_pessimistic_lock` or `prewrite` to get the correct value (which is what the comment in L272 means). Then, only the first time a key is written will be slow. I think this way is a bit better.

Comment (sticnarf) [reply components/txn_types/src/write.rs:276]: > By the way, GC may change the real number of versions to last change (by deleting old `Lock`s and `Rollback`s). Does that matter?

Right... This also brings inaccuracy. I don't have a good idea now but I don't think it's a very serious problem. The performance just doesn't come back to optimism, and it is likely for such a key to pile up enough versions again.

Anyway, it's always pleasant to solve this problem if possible. Do you have any good idea...?

Comment (MyonKeminta) [reply components/txn_types/src/write.rs:276]: > > By the way, GC may change the real number of versions to last change (by deleting old `Lock`s and `Rollback`s). Does that matter?
> 
> Right... This also brings inaccuracy. I don't have a good idea now but I don't think it's a very serious problem. The performance just doesn't come back to optimism, and it is likely for such a key to pile up enough versions again.
> 
> Anyway, it's always pleasant to solve this problem if possible. Do you have any good idea...?

I have no other idea either. I think it's ok to ignore the inaccuracy too. (I'm just confirming that you already noticed this case

**Pull request merged**